src/lib/pickles/test/ec-scaling-foreign/test.ml

```diff
@@ -0,0 +1,345 @@
+    let x = sub_chain ~range_check s_squared_sub_a_x b_x in
+    (* x_change = b_x - x *)
+    let neg_x_change = sub_chain ~range_check x b_x in
+    (* y_change = (b_x - x) * s *)
+    let neg_y_change = mul_chain ~range_check neg_x_change s in
+    let y = exists foreign_typ ~compute:(fun () -> ()) in
+    (* y = (b_x - x) * s - b_y *)
+    let neg_b_y = add_chain ~range_check neg_y_change y in
+    let zero' = add_chain ~range_check neg_b_y b_y in
+    assert_equal zero zero' ;
+    (* Run the deferred range checks *)
+    List.iter ~f:actual_range_check !range_checks ;
+    (x, y)
+
+  (*
+     x = s^2 - 2 A_x
+     y = - A_y + s (A_x - x)
+
+     s = (3 A_x^2 + a) / (2 A_y)
+  *)
+  let ec_double a (a_x, a_y) =
+    let actual_range_check = range_check in
+    let range_checks = ref [] in
+    let range_check x = range_checks := x :: !range_checks in
+    (* a_x_squared = a_x * a_x *)
+    let a_x_squared = mul_chain ~range_check a_x a_x in
+    (* two_a_x_squared = 2 * a_x * a_x *)
+    let two_a_x_squared = add_chain ~range_check a_x_squared a_x_squared in
+    (* three_a_x_squared = 3 * a_x * a_x *)
+    let three_a_x_squared =
+      add_chain ~range_check two_a_x_squared a_x_squared
+    in
+    (* s_numerator = 3 * a_x * a_x + a *)
+    let s_numerator = add ~range_check three_a_x_squared a in
+    let s_denominator = add_chain ~range_check a_y a_y in
+    let s = exists foreign_typ ~compute:(fun () -> ()) in
+    let s_numerator' = mul ~range_check s_denominator s in
+    assert_equal s_numerator s_numerator' ;
+    let s_squared = mul_chain ~range_check s s in
+    let s_squared_sub_a_x = sub_chain ~range_check s_squared a_x in
+    let x = sub_chain ~range_check s_squared_sub_a_x a_x in
+    let x_sub_a_x = sub_chain ~range_check x a_x in
+    let s_mul_a_sub_a_x = mul_chain ~range_check x_sub_a_x s in
+    let neg_y = add_chain ~range_check s_mul_a_sub_a_x a_y in
+    let y = exists foreign_typ ~compute:(fun () -> ()) in
+    let zero' = add_chain ~range_check neg_y y in
+    assert_equal zero zero' ;
+    (* Run the deferred range checks *)
+    List.iter ~f:actual_range_check !range_checks ;
+    (x, y)
+
+  let pt_typ = Typ.tuple2 foreign_typ foreign_typ
+
+  let circuit a pt scalar_bits =
+    let res, _is_zero, _pt =
+      Array.fold
+        ~init:((zero, zero), Boolean.true_, pt)
+        scalar_bits
+        ~f:(fun (acc, is_zero, pt) bit ->
+          let still_zero = Boolean.(is_zero &&& not bit) in
+          let acc_base = if_ is_zero ~typ:pt_typ ~then_:pt ~else_:acc in
+          let acc_res = ec_add pt acc_base in
+          let acc =
+            if_ bit ~typ:pt_typ
+              ~then_:(if_ is_zero ~typ:pt_typ ~then_:pt ~else_:acc_res)
+              ~else_:acc
+          in
+          (* NB: We do an unnecessary final double. Optimise before using anywhere. *)
+          let double_pt = ec_double a pt in
+          (acc, still_zero, double_pt) )
+    in
+    ignore (res : foreign * foreign)
```

Comment:
Do/will we care about `res` ? 

If not, I'd favor the more explicit (same type annotations):
```suggestion
ignore
      ( Array.fold
          ~init:((zero, zero), Boolean.true_, pt)
          scalar_bits
          ~f:(fun (acc, is_zero, pt) bit ->
            let still_zero = Boolean.(is_zero &&& not bit) in
            let acc_base = if_ is_zero ~typ:pt_typ ~then_:pt ~else_:acc in
            let acc_res = ec_add pt acc_base in
            let acc =
              if_ bit ~typ:pt_typ
                ~then_:(if_ is_zero ~typ:pt_typ ~then_:pt ~else_:acc_res)
                ~else_:acc
            in
            (* NB: We do an unnecessary final double. Optimise before using anywhere. *)
            let double_pt = ec_double a pt in
            (acc, still_zero, double_pt) )
        : (foreign * foreign) * _ * _ )
```